src/ucp/api/ucp.h

```diff
@@ -916,6 +959,40 @@ typedef struct ucp_mem_map_params {
      unsigned               flags;
 } ucp_mem_map_params_t;
 
+enum ucp_group_rank_distance {
```

Comment:
Is the assumption here that MPI (or whoever using it) has some info about locality ? 
If so, maybe this is something that can be passed (optionally) during endpoint creation, instead of group ? It would make sense to cache on EP only if EP code actually can benefit from this info in p2p flow.

Reply:
This info would be helpful, but not mandatory. It will work if you pass the same distance for all peer ranks. It does (currently) assume the information is consistent among peers: distance(A,B) == distance(B,A).

The reason I put it here is that it makes UCP responsible for EP creation - it chooses which EPs to create ("on-demand") based on the group layout. For example - in a tree broadcast a "node-leader" will connect to all his host-based peers, but only a couple of out-of-host peers.

Reply:
Makes sense to me. We would have to generalize this a bit, since we also care about GPUs. Essentially you have shared memory domain and external (non-coherent) domain. Within shared memory domain you will have NUMA, socket, etc. Within non-coherent domain you also may have levels of hierarchy (TOR switches , etc.)